File: 2018summer/07.12/solutions_2015/bountyhii/submissions/accepted/BountyHunterHKEgor.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.TreeSet;


public class BountyHunterHKEgor {

	

	public static void main(String[] args)throws Exception{
        new BountyHunterHKEgor().run();
    }

	



	@SuppressWarnings("unchecked")
	private void run() throws Exception{

		  BufferedReader reader=new BufferedReader(new InputStreamReader(System.in));


		  int N=Integer.parseInt(reader.readLine());	  

		  int graphSize=2*N+2;
		  Set<Integer>[] graph=new Set[graphSize];

		  for(int i=0;i<graphSize;i++)
			graph[i]=new TreeSet<Integer>();

			

		  for(int i=0;i<N;i++){
			  graph[graphSize-2].add(i);
			  graph[N+i].add(graphSize-1);
		  }


		  for(int i=0;i<N;i++){

			  String[] goals=reader.readLine().split(" ");
			  
			  for(int j=1;j<goals.length;j++){
				  int currentGoal=Integer.parseInt(goals[j]);
				  graph[i].add(N+currentGoal);
			  }
		  }

		  HopcroftKarp hk=new HopcroftKarp();

		  System.out.println(N-hk.matchingHopcroftKarp(graph, graphSize-2, graphSize-1));
	}



	public class HopcroftKarp {

	    
		
		 Set<Integer>[] getBfsTree(Set<Integer>[] graph, int source, int target){
			int n=graph.length;
			int[] used=new int[n];
			Arrays.fill(used, -1);
			
			@SuppressWarnings("unchecked")
			Set<Integer>[] result=new Set[n];
			for(int i=0;i<n;result[i++]=new TreeSet<Integer>());
			Queue<Integer> q=new LinkedList<Integer>();
			
			used[source]=0;
			q.offer(source);
			while(!q.isEmpty()){
				int current=q.poll();
				if(current==target)
					return result;
				
				int currentDist=used[current];
				for(int next:graph[current]){
					if(used[next]==-1){
						used[next]=currentDist+1;
						q.offer(next);
					}
					if(used[next]==currentDist+1)
						result[current].add(next);
				}
			}
			//no path to target:
			return null;
		}
		
		 boolean dfs(Set<Integer>[] graph, Set<Integer>[] bfsTree, int source,  int target, boolean[] used){
			if(source==target)
				return true;
			
			used[source]=true;
			for(int next: bfsTree[source]){
				if(!used[next] && dfs(graph, bfsTree, next, target, used)){
					graph[source].remove(next);
					graph[next].add(source);
					return true;
				}
			}
			return false;
		}
	
		
		//matching with hopcroft-karp O(m*sqrt(n))
		int matchingHopcroftKarp(Set<Integer>[] graph, int source, int target){
			int result=0;
			int n=graph.length;
			boolean[] used=new boolean[n];
			while(true){
				
				Set<Integer>[] bfsTree=getBfsTree(graph, source, target);
				if(bfsTree==null) 
					return result;

				
				Arrays.fill(used, false);
				while(dfs(graph, bfsTree, source, target, used))
					result++;
				
			}
			
		}
	}
}

File: 2018summer/07.12/solutions_2015/bountyhii/submissions/accepted/BountyHunterHKEgor.py
import sys
from collections import deque

def run():
    N=int(sys.stdin.readline().strip())

    graphSize=2*N+2;
    graph = [set() for x in range(graphSize)]

    for i in range(N):
        graph[graphSize-2].add(i)
        graph[N+i].add(graphSize-1)

    for i in range(N):
        goals = sys.stdin.readline().split()
        for j in range(1, len(goals)):
            currentGoal=int(goals[j])
            graph[i].add(N+currentGoal)

    hk = HopcroftKarp()

    print(N-hk.matchingHopcroftKarp(graph, graphSize-2, graphSize-1))


class HopcroftKarp(object):

    def getBfsTree(self, graph, source, target):
        n=len(graph)
        used= [-1 for x in range(n)]
            
        result = [set() for x in range(n)]
        
        q= deque()
            
        used[source]=0
        q.append(source)
        
        while q:
            current=q.pop()
            if(current==target):
                return result
                    
            currentDist=used[current]
            for _next in graph[current]:
                if used[_next] == -1:
                    used[_next] = currentDist+1
                    q.append(_next)
                    
                if used[_next]==currentDist+1:
                    result[current].add(_next)
            
            
            #//no path to target:
        return None
		
    def dfs(self, graph, bfsTree, source,  target, used):
        if source==target:
            return True
			
        used[source]=True
        for _next in bfsTree[source]:
            if not used[_next] and self.dfs(graph, bfsTree, _next, target, used):
                graph[source].remove(_next)
                graph[_next].add(source)
                return True
            
        
        return False
	
		
    #//matching with hopcroft-karp O(m*sqrt(n))
    def matchingHopcroftKarp(self, graph, source, target):
        result=0
        n=len(graph)
        used=[False for x in range(n)]
        while True:
            
            bfsTree=self.getBfsTree(graph, source, target);
            if bfsTree==None:
                return result

            for x in range(n):
                used[x] = False
            while self.dfs(graph, bfsTree, source, target, used):
                result+=1
				

run()
